File: Python/Problems/Problems/problem67.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem67(Problem):
    def test(self) -> bool:
        a = "1010"
        b = "1011"

        expected = "10101"

        result = self.addBinary(a, b)

        print(result)

        return result == expected

    def addBinary(self, a: str, b: str) -> str:
        n = len(a)
        m = len(b)
        big = max(n, m)
        result = ['0'] * (big + 1)
        carry = 0
        for i in range(big + 1):
            first = ord(a[n - i - 1]) - ord('0') if  i < n else 0
            second = ord(b[m - i - 1]) - ord('0') if i < m else 0
            total = carry + first + second
            result[big - i] = chr(total % 2 + ord('0'))
            carry = 1 if total > 1 else 0
        resultStr = ''.join(result)
        return resultStr[1:] if resultStr[0] == '0' else resultStr

File: C++/Problems/Problems/problem_67.cpp
#include "problem_67.h"

bool problem_67::test() {
    std::string a = "1010";
    std::string b = "1011";

    std::string expected = "10101";

    auto result = addBinary(a, b);

    std::cout << result << std::endl;

    return result == expected;
}

std::string problem_67::addBinary(std::string a, std::string b) {
    int n = a.size();
    int m = b.size();
    int big = std::max(n, m);
    std::string result(big + 1, '0');
    int carry = 0;
    for (int i = 0; i <= big; ++i)
    {
        int first = i < n ? a[n - i - 1] - '0' : 0;
        int second = i < m ? b[m - i - 1] - '0' : 0;
        int total = carry + first + second;
        result[big - i] = (char)(total % 2 + '0');
        carry = total > 1 ? 1 : 0;
    }

    return result[0] == '0' ? result.substr(1) : result;
}
